**Hello, welcome to the preprocessing notebook of our data science project: predicting successful startup!**

We will try to include as much details as possible with markdowns :)

# Data Import

In [383]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [384]:
data = pd.read_csv('/Users/danyhersco/startups.csv')

# Data Exploration

In [385]:
data.shape

(314486, 19)

Let's quickly drop real duplicates (see below for "false duplicates"):

In [386]:
data = data.drop_duplicates()
data.shape

(312667, 19)

We extracted the dataset from DBeaver, after running an SQL query. As a result of multiple `LEFT JOIN` to include `industry` and `technology` features, many startups have duplicates.

Consequently, let's check for the number of unique values for each feature of our dataframe.

In [387]:
data.describe(include='all')

id                 name  \
count                                 312667               312667   
unique                                 56667                55665   
top     4f97ceac-342b-425d-a947-f03c48012b20  Arthur Intelligence   
freq                                     216                  216   
mean                                     NaN                  NaN   
std                                      NaN                  NaN   
min                                      NaN                  NaN   
25%                                      NaN                  NaN   
50%                                      NaN                  NaN   
75%                                      NaN                  NaN   
max                                      NaN                  NaN   

                         website              short_description ipo_status  \
count                     309473                         312664     312667   
unique                     54614                          56142          3   
top     https://www.goarthur.ai/  Limited information available    private   
freq                         216                            597     307633   
mean                         NaN                            NaN        NaN   
std                          NaN                            NaN        NaN   
min                          NaN                            NaN        NaN   
25%                          NaN                            NaN        NaN   
50%                          NaN                            NaN        NaN   
75%                          NaN                            NaN        NaN   
max                          NaN                            NaN        NaN   

        founded_on went_public_on   exited_on  num_funding_rounds  \
count       312667           5360       21964       312667.000000   
unique        1083            507        1370                 NaN   
top     2016-01-01     2021-11-11  2021-02-19                 NaN   
freq         67127            114         146                 NaN   
mean           NaN            NaN         NaN            3.406656   
std            NaN            NaN         NaN            2.419148   
min            NaN            NaN         NaN            1.000000   
25%            NaN            NaN         NaN            2.000000   
50%            NaN            NaN         NaN            3.000000   
75%            NaN            NaN         NaN            5.000000   
max            NaN            NaN         NaN           22.000000   

       last_equity_funding_type  \
count                    287213   
unique                       18   
top                        seed   
freq                     107046   
mean                        NaN   
std                         NaN   
min                         NaN   
25%                         NaN   
50%                         NaN   
75%                         NaN   
max                         NaN   

                                last_equity_funding_total last_funding_at  \
count                                              235134          312667   
unique                                              15347            2905   
top     {"currency":"USD","amount":100000000,"amountUS...      2019-01-01   
freq                                                 4766            2877   
mean                                                  NaN             NaN   
std                                                   NaN             NaN   
min                                                   NaN             NaN   
25%                                                   NaN             NaN   
50%                                                   NaN             NaN   
75%                                                   NaN             NaN   
max                                                   NaN             NaN   

       headquartersCountry headquartersRegion  employeeCount industry_name  \
count               311

We have **55665** unique startups in our dataset. The next step will be to remove all duplicates while treating the non-duplicated columns.

Here is how our dataset look like:

In [388]:
data.head()

id          name  \
0  75139f4c-6d9d-4a03-8b72-61f73f7be74a  Youcan Robot   
1  75139f4c-6d9d-4a03-8b72-61f73f7be74a  Youcan Robot   
2  75139f4c-6d9d-4a03-8b72-61f73f7be74a  Youcan Robot   
3  75139f4c-6d9d-4a03-8b72-61f73f7be74a  Youcan Robot   
4  75139f4c-6d9d-4a03-8b72-61f73f7be74a  Youcan Robot   

                        website  \
0  https://www.youcanrobot.com/   
1  https://www.youcanrobot.com/   
2  https://www.youcanrobot.com/   
3  https://www.youcanrobot.com/   
4  https://www.youcanrobot.com/   

                                   short_description ipo_status  founded_on  \
0  Youcan Robot is an integrated company in the w...    private  2016-01-01   
1  Youcan Robot is an integrated company in the w...    private  2016-01-01   
2  Youcan Robot is an integrated company in the w...    private  2016-01-01   
3  Youcan Robot is an integrated company in the w...    private  2016-01-01   
4  Youcan Robot is an integrated company in the w...    private  2016-01-01   

  went_public_on exited_on  num_funding_rounds last_equity_funding_type  \
0            NaN       NaN                   1                     seed   
1            NaN       NaN                   1                     seed   
2            NaN       NaN                   1                     seed   
3            NaN       NaN                   1                     seed   
4            NaN       NaN                   1                     seed   

  last_equity_funding_total last_funding_at headquartersCountry  \
0                       NaN      2018-05-30                  CN   
1                       NaN      2018-05-30                  CN   
2                       NaN      2018-05-30                  CN   
3                       NaN      2018-05-30                  CN   
4                       NaN      2018-05-30                  CN   

  headquartersRegion  employeeCount            industry_name  \
0           Shanghai           31.0                    Video   
1           Shanghai           31.0                    Video   
2           Shanghai           31.0        Computer Hardware   
3           Shanghai           31.0        Computer Hardware   
4           Shanghai           31.0  Media and Entertainment   

           technology_name              announcedOn moneyRaised  
0  Science and Engineering  2018-05-30 00:00:00.000         NaN  
1                 Hardware  2018-05-30 00:00:00.000         NaN  
2  Science and Engineering  2018-05-30 00:00:00.000         NaN  
3                 Hardware  2018-05-30 00:00:00.000         NaN  
4  Science and Engineering  2018-05-30 00:00:00.000         NaN

# Remove the duplicates through encoding

This problem will be challenging as there is not really fully duplicated rows. As we said above, it is just a result of the `LEFT JOIN` that multiplicated the rows for one company (several industries, several technologies, and several funding rounds). We have to one hot encode the latter features, which is what we will be doing in this section.

At the end of this section, we should reach a shape of `(55665, >18)`.

In [389]:
data.head()

id          name  \
0  75139f4c-6d9d-4a03-8b72-61f73f7be74a  Youcan Robot   
1  75139f4c-6d9d-4a03-8b72-61f73f7be74a  Youcan Robot   
2  75139f4c-6d9d-4a03-8b72-61f73f7be74a  Youcan Robot   
3  75139f4c-6d9d-4a03-8b72-61f73f7be74a  Youcan Robot   
4  75139f4c-6d9d-4a03-8b72-61f73f7be74a  Youcan Robot   

                        website  \
0  https://www.youcanrobot.com/   
1  https://www.youcanrobot.com/   
2  https://www.youcanrobot.com/   
3  https://www.youcanrobot.com/   
4  https://www.youcanrobot.com/   

                                   short_description ipo_status  founded_on  \
0  Youcan Robot is an integrated company in the w...    private  2016-01-01   
1  Youcan Robot is an integrated company in the w...    private  2016-01-01   
2  Youcan Robot is an integrated company in the w...    private  2016-01-01   
3  Youcan Robot is an integrated company in the w...    private  2016-01-01   
4  Youcan Robot is an integrated company in the w...    private  2016-01-01   

  went_public_on exited_on  num_funding_rounds last_equity_funding_type  \
0            NaN       NaN                   1                     seed   
1            NaN       NaN                   1                     seed   
2            NaN       NaN                   1                     seed   
3            NaN       NaN                   1                     seed   
4            NaN       NaN                   1                     seed   

  last_equity_funding_total last_funding_at headquartersCountry  \
0                       NaN      2018-05-30                  CN   
1                       NaN      2018-05-30                  CN   
2                       NaN      2018-05-30                  CN   
3                       NaN      2018-05-30                  CN   
4                       NaN      2018-05-30                  CN   

  headquartersRegion  employeeCount            industry_name  \
0           Shanghai           31.0                    Video   
1           Shanghai           31.0                    Video   
2           Shanghai           31.0        Computer Hardware   
3           Shanghai           31.0        Computer Hardware   
4           Shanghai           31.0  Media and Entertainment   

           technology_name              announcedOn moneyRaised  
0  Science and Engineering  2018-05-30 00:00:00.000         NaN  
1                 Hardware  2018-05-30 00:00:00.000         NaN  
2  Science and Engineering  2018-05-30 00:00:00.000         NaN  
3                 Hardware  2018-05-30 00:00:00.000         NaN  
4  Science and Engineering  2018-05-30 00:00:00.000         NaN

##  `industry` encoding

In [442]:
data_ind = data[['id', 'industry_name']].drop_duplicates()
data_ind.head()

id            industry_name
0  75139f4c-6d9d-4a03-8b72-61f73f7be74a                    Video
2  75139f4c-6d9d-4a03-8b72-61f73f7be74a        Computer Hardware
4  75139f4c-6d9d-4a03-8b72-61f73f7be74a  Media and Entertainment
6  75139f4c-6d9d-4a03-8b72-61f73f7be74a     Consumer Electronics
8  441e5d33-140e-40c6-ab5d-35f65ab8feed     Privacy and Security

In [443]:
from sklearn.preprocessing import OneHotEncoder

encoder = OneHotEncoder(sparse=False)
data_ind[encoder.get_feature_names_out()] = encoder.fit_transform(data_ind[['industry_name']])

In [444]:
data_ind.head()

id            industry_name  \
0  75139f4c-6d9d-4a03-8b72-61f73f7be74a                    Video   
2  75139f4c-6d9d-4a03-8b72-61f73f7be74a        Computer Hardware   
4  75139f4c-6d9d-4a03-8b72-61f73f7be74a  Media and Entertainment   
6  75139f4c-6d9d-4a03-8b72-61f73f7be74a     Consumer Electronics   
8  441e5d33-140e-40c6-ab5d-35f65ab8feed     Privacy and Security   

   industry_name_Advertising  industry_name_Agriculture and Farming  \
0                        0.0                                    0.0   
2                        0.0                                    0.0   
4                        0.0                                    0.0   
6                        0.0                                    0.0   
8                        0.0                                    0.0   

   industry_name_Clothing and Apparel  industry_name_Commerce and Shopping  \
0                                 0.0                                  0.0   
2                                 0.0                                  0.0   
4                                 0.0                                  0.0   
6                                 0.0                                  0.0   
8                                 0.0                                  0.0   

   industry_name_Community and Lifestyle  industry_name_Computer Hardware  \
0                                    0.0                              0.0   
2                                    0.0                              1.0   
4                                    0.0                              0.0   
6                                    0.0                              0.0   
8                                    0.0                              0.0   

   industry_name_Consumer Electronics  industry_name_Consumer Goods  ...  \
0                                 0.0                           0.0  ...   
2                                 0.0                           0.0  ...   
4                                 0.0                           0.0  ...   
6                                 1.0                           0.0  ...   
8                                 0.0                           0.0  ...   

   industry_name_Privacy and Security  industry_name_Professional Services  \
0                                 0.0                                  0.0   
2                                 0.0                                  0.0   
4                                 0.0                                  0.0   
6                                 0.0                                  0.0   
8                                 1.0                                  0.0   

   industry_name_Real Estate and Construction  \
0                                         0.0   
2                                         0.0   
4                                         0.0   
6                                         0.0   
8                                         0.0   

   industry_name_Sales and Marketing  industry_name_Software  \
0                                0.0                     0.0   
2                                0.0                     0.0   
4                                0.0                     0.0   
6                                0.0                     0.0   
8                                0.0                     0.0   

   industry_name_Sports  industry_name_Transportation  \
0                   0.0                           0.0   
2                   0.0                           0.0   
4                   0.0                           0.0   
6                   0.0                           0.0   
8                   0.0                           0.0   

   industry_name_Travel and Tourism  industry_name_Video  industry_name_nan  
0                               0.0                  1.0                0.0  
2                               0.0                  0.0                0.0  
4                               0.0                  0.0                0.0  
6       

In [445]:
for column in data_ind.columns[2:]:
    data_ind.rename(columns={column: column[14:]}, inplace=True)

In [446]:
data_ind.head()

id            industry_name  Advertising  \
0  75139f4c-6d9d-4a03-8b72-61f73f7be74a                    Video          0.0   
2  75139f4c-6d9d-4a03-8b72-61f73f7be74a        Computer Hardware          0.0   
4  75139f4c-6d9d-4a03-8b72-61f73f7be74a  Media and Entertainment          0.0   
6  75139f4c-6d9d-4a03-8b72-61f73f7be74a     Consumer Electronics          0.0   
8  441e5d33-140e-40c6-ab5d-35f65ab8feed     Privacy and Security          0.0   

   Agriculture and Farming  Clothing and Apparel  Commerce and Shopping  \
0                      0.0                   0.0                    0.0   
2                      0.0                   0.0                    0.0   
4                      0.0                   0.0                    0.0   
6                      0.0                   0.0                    0.0   
8                      0.0                   0.0                    0.0   

   Community and Lifestyle  Computer Hardware  Consumer Electronics  \
0                      0.0                0.0                   0.0   
2                      0.0                1.0                   0.0   
4                      0.0                0.0                   0.0   
6                      0.0                0.0                   1.0   
8                      0.0                0.0                   0.0   

   Consumer Goods  ...  Privacy and Security  Professional Services  \
0             0.0  ...                   0.0                    0.0   
2             0.0  ...                   0.0                    0.0   
4             0.0  ...                   0.0                    0.0   
6             0.0  ...                   0.0                    0.0   
8             0.0  ...                   1.0                    0.0   

   Real Estate and Construction  Sales and Marketing  Software  Sports  \
0                           0.0                  0.0       0.0     0.0   
2                           0.0                  0.0       0.0     0.0   
4                           0.0                  0.0       0.0     0.0   
6                           0.0                  0.0       0.0     0.0   
8                           0.0                  0.0       0.0     0.0   

   Transportation  Travel and Tourism  Video  nan  
0             0.0                 0.0    1.0  0.0  
2             0.0                 0.0    0.0  0.0  
4             0.0                 0.0    0.0  0.0  
6             0.0                 0.0    0.0  0.0  
8             0.0                 0.0    0.0  0.0  

[5 rows x 43 columns]

In [447]:
data_ind = data_ind.groupby('id').max()
data_ind.head()

industry_name  \
id                                                                     
00003f34-ac89-4090-b592-30f2b70ff5a4  Environment and Sustainability   
00014e45-238c-4ce1-a9d1-de7c6d5bbf2c         Community and Lifestyle   
0001b1aa-4a8b-4ee2-b4c0-6bce0318c2c1                     Health Care   
0003088d-8c0b-46f6-910a-6dea65777ac4                  Transportation   
0004d27e-6d98-4f29-a708-b989e3e38f81                     Health Care   

                                      Advertising  Agriculture and Farming  \
id                                                                           
00003f34-ac89-4090-b592-30f2b70ff5a4          0.0                      0.0   
00014e45-238c-4ce1-a9d1-de7c6d5bbf2c          0.0                      0.0   
0001b1aa-4a8b-4ee2-b4c0-6bce0318c2c1          0.0                      0.0   
0003088d-8c0b-46f6-910a-6dea65777ac4          0.0                      0.0   
0004d27e-6d98-4f29-a708-b989e3e38f81          0.0                      0.0   

                                      Clothing and Apparel  \
id                                                           
00003f34-ac89-4090-b592-30f2b70ff5a4                   0.0   
00014e45-238c-4ce1-a9d1-de7c6d5bbf2c                   0.0   
0001b1aa-4a8b-4ee2-b4c0-6bce0318c2c1                   0.0   
0003088d-8c0b-46f6-910a-6dea65777ac4                   0.0   
0004d27e-6d98-4f29-a708-b989e3e38f81                   0.0   

                                      Commerce and Shopping  \
id                                                            
00003f34-ac89-4090-b592-30f2b70ff5a4                    0.0   
00014e45-238c-4ce1-a9d1-de7c6d5bbf2c                    0.0   
0001b1aa-4a8b-4ee2-b4c0-6bce0318c2c1                    0.0   
0003088d-8c0b-46f6-910a-6dea65777ac4                    0.0   
0004d27e-6d98-4f29-a708-b989e3e38f81                    0.0   

                                      Community and Lifestyle  \
id                                                              
00003f34-ac89-4090-b592-30f2b70ff5a4                      0.0   
00014e45-238c-4ce1-a9d1-de7c6d5bbf2c                      1.0   
0001b1aa-4a8b-4ee2-b4c0-6bce0318c2c1                      0.0   
0003088d-8c0b-46f6-910a-6dea65777ac4                      0.0   
0004d27e-6d98-4f29-a708-b989e3e38f81                      0.0   

                                      Computer Hardware  Consumer Electronics  \
id                                                                              
00003f34-ac89-4090-b592-30f2b70ff5a4                0.0                   0.0   
00014e45-238c-4ce1-a9d1-de7c6d5bbf2c                0.0                   0.0   
0001b1aa-4a8b-4ee2-b4c0-6bce0318c2c1                0.0                   0.0   
0003088d-8c0b-46f6-910a-6dea65777ac4                0.0                   0.0   
0004d27e-6d98-4f29-a708-b989e3e38f81                0.0                   0.0   

                                      Consumer Goods  Content and Publishing  \
id                                                                             
00003f34-ac89-4090-b592-30f2b70ff5a4             0.0                     0.0   
00014e45-238c-4ce1-a9d1-de7c6d5bbf2c             0.0                     0.0   
0001b1aa-4a8b-4ee2-b4c0-6bce0318c2c1             0.0                     0.0   
0003088d-8c0b-46f6-910a-6dea65777ac4             0.0                     0.0   
0004d27e-6d98-4f29-a708-b989e3e38f81             0.0                     0.0   

                                      ...  Privacy and Security  \
id                                    ...                         
00003f34-ac89-4090-b592-30f2b70ff5a4  ...                   0.0   
00014e45-238c-4ce1-a9d1-de7c6d5bbf2c  ...                   0.0   
0001b1aa-4a8b-4ee2-b4c0-6bce0318c2c1  ...                   0.0   
0003088d-8c0b-46f6-910a-6dea65777ac4  ...                   0.0   
0004d27e-6d98-4f29-a708-b989e3e38f81  ...                   0.0   

                                      Professional Services  \
id      

In [448]:
data_ind.reset_index(inplace=True)
data_ind.drop(columns=['nan', 'industry_name'], inplace=True)

In [449]:
data_ind.shape

(56667, 41)

##  `technology` encoding

In [453]:
data_tec = data[['id', 'technology_name']].drop_duplicates()
data_tec.head()

id          technology_name
0   75139f4c-6d9d-4a03-8b72-61f73f7be74a  Science and Engineering
1   75139f4c-6d9d-4a03-8b72-61f73f7be74a                 Hardware
8   441e5d33-140e-40c6-ab5d-35f65ab8feed                 Software
10  59b8a226-eb24-4e7b-851d-924e0385d01c                 Software
14  f786e92e-7ac1-4035-9644-de30b8d8b6a5            Biotechnology

In [454]:
encoder2 = OneHotEncoder(sparse=False)
data_tec[encoder2.get_feature_names_out()] = encoder2.fit_transform(data_tec[['technology_name']])

In [455]:
data_tec.head()

id          technology_name  \
0   75139f4c-6d9d-4a03-8b72-61f73f7be74a  Science and Engineering   
1   75139f4c-6d9d-4a03-8b72-61f73f7be74a                 Hardware   
8   441e5d33-140e-40c6-ab5d-35f65ab8feed                 Software   
10  59b8a226-eb24-4e7b-851d-924e0385d01c                 Software   
14  f786e92e-7ac1-4035-9644-de30b8d8b6a5            Biotechnology   

    technology_name_AR and VR  technology_name_Artificial Intelligence  \
0                         0.0                                      0.0   
1                         0.0                                      0.0   
8                         0.0                                      0.0   
10                        0.0                                      0.0   
14                        0.0                                      0.0   

    technology_name_Biotechnology  technology_name_BlockChain  \
0                             0.0                         0.0   
1                             0.0                         0.0   
8                             0.0                         0.0   
10                            0.0                         0.0   
14                            1.0                         0.0   

    technology_name_Hardware  technology_name_Science and Engineering  \
0                        0.0                                      1.0   
1                        1.0                                      0.0   
8                        0.0                                      0.0   
10                       0.0                                      0.0   
14                       0.0                                      0.0   

    technology_name_Software  technology_name_Sustainability  \
0                        0.0                             0.0   
1                        0.0                             0.0   
8                        1.0                             0.0   
10                       1.0                             0.0   
14                       0.0                             0.0   

    technology_name_nan  
0                   0.0  
1                   0.0  
8                   0.0  
10                  0.0  
14                  0.0

In [456]:
for column in data_tec.columns[2:]:
    data_tec.rename(columns={column: column[16:]}, inplace=True)

In [457]:
data_tec.head()

id          technology_name  AR and VR  \
0   75139f4c-6d9d-4a03-8b72-61f73f7be74a  Science and Engineering        0.0   
1   75139f4c-6d9d-4a03-8b72-61f73f7be74a                 Hardware        0.0   
8   441e5d33-140e-40c6-ab5d-35f65ab8feed                 Software        0.0   
10  59b8a226-eb24-4e7b-851d-924e0385d01c                 Software        0.0   
14  f786e92e-7ac1-4035-9644-de30b8d8b6a5            Biotechnology        0.0   

    Artificial Intelligence  Biotechnology  BlockChain  Hardware  \
0                       0.0            0.0         0.0       0.0   
1                       0.0            0.0         0.0       1.0   
8                       0.0            0.0         0.0       0.0   
10                      0.0            0.0         0.0       0.0   
14                      0.0            1.0         0.0       0.0   

    Science and Engineering  Software  Sustainability  nan  
0                       1.0       0.0             0.0  0.0  
1                       0.0       0.0             0.0  0.0  
8                       0.0       1.0             0.0  0.0  
10                      0.0       1.0             0.0  0.0  
14                      0.0       0.0             0.0  0.0

In [458]:
data_tec = data_tec.groupby('id').max()
data_tec.head()

technology_name  AR and VR  \
id                                                                
00003f34-ac89-4090-b592-30f2b70ff5a4  Sustainability        0.0   
00014e45-238c-4ce1-a9d1-de7c6d5bbf2c        Software        0.0   
0001b1aa-4a8b-4ee2-b4c0-6bce0318c2c1        Software        0.0   
0003088d-8c0b-46f6-910a-6dea65777ac4        Software        0.0   
0004d27e-6d98-4f29-a708-b989e3e38f81             NaN        0.0   

                                      Artificial Intelligence  Biotechnology  \
id                                                                             
00003f34-ac89-4090-b592-30f2b70ff5a4                      0.0            0.0   
00014e45-238c-4ce1-a9d1-de7c6d5bbf2c                      0.0            0.0   
0001b1aa-4a8b-4ee2-b4c0-6bce0318c2c1                      0.0            0.0   
0003088d-8c0b-46f6-910a-6dea65777ac4                      0.0            0.0   
0004d27e-6d98-4f29-a708-b989e3e38f81                      0.0            0.0   

                                      BlockChain  Hardware  \
id                                                           
00003f34-ac89-4090-b592-30f2b70ff5a4         0.0       0.0   
00014e45-238c-4ce1-a9d1-de7c6d5bbf2c         0.0       0.0   
0001b1aa-4a8b-4ee2-b4c0-6bce0318c2c1         0.0       0.0   
0003088d-8c0b-46f6-910a-6dea65777ac4         0.0       0.0   
0004d27e-6d98-4f29-a708-b989e3e38f81         0.0       0.0   

                                      Science and Engineering  Software  \
id                                                                        
00003f34-ac89-4090-b592-30f2b70ff5a4                      0.0       0.0   
00014e45-238c-4ce1-a9d1-de7c6d5bbf2c                      0.0       1.0   
0001b1aa-4a8b-4ee2-b4c0-6bce0318c2c1                      0.0       1.0   
0003088d-8c0b-46f6-910a-6dea65777ac4                      0.0       1.0   
0004d27e-6d98-4f29-a708-b989e3e38f81                      0.0       0.0   

                                      Sustainability  nan  
id                                                         
00003f34-ac89-4090-b592-30f2b70ff5a4             1.0  0.0  
00014e45-238c-4ce1-a9d1-de7c6d5bbf2c             0.0  0.0  
0001b1aa-4a8b-4ee2-b4c0-6bce0318c2c1             0.0  0.0  
0003088d-8c0b-46f6-910a-6dea65777ac4             0.0  0.0  
0004d27e-6d98-4f29-a708-b989e3e38f81             0.0  1.0

In [459]:
data_tec.reset_index(inplace=True)
data_tec.drop(columns=['nan', 'technology_name'], inplace=True)

In [460]:
data_tec.shape

(56667, 9)

## `funding_round` encoding

In [418]:
data.columns

Index(['id', 'name', 'website', 'short_description', 'ipo_status',
       'founded_on', 'went_public_on', 'exited_on', 'num_funding_rounds',
       'last_equity_funding_type', 'last_equity_funding_total',
       'last_funding_at', 'headquartersCountry', 'headquartersRegion',
       'employeeCount', 'industry_name', 'technology_name', 'announcedOn',
       'moneyRaised'],
      dtype='object')

In [419]:
data_fun = data[['id', 'announcedOn', 'moneyRaised']].drop_duplicates()

In [420]:
data_fun.head()

id              announcedOn  \
0   75139f4c-6d9d-4a03-8b72-61f73f7be74a  2018-05-30 00:00:00.000   
8   441e5d33-140e-40c6-ab5d-35f65ab8feed  2022-11-30 00:00:00.000   
9   441e5d33-140e-40c6-ab5d-35f65ab8feed  2021-06-17 00:00:00.000   
10  59b8a226-eb24-4e7b-851d-924e0385d01c  2016-07-11 00:00:00.000   
14  f786e92e-7ac1-4035-9644-de30b8d8b6a5  2018-08-01 00:00:00.000   

                                          moneyRaised  
0                                                 NaN  
8   {"amount": 5000000, "currency": "USD", "amount...  
9                                                 NaN  
10  {"amount": 50000000, "currency": "USD", "amoun...  
14                                                NaN

In [421]:
data_fun['announcedOn'] = pd.to_datetime(data_fun['announcedOn'])
data_fun.sort_values(by=['id', 'announcedOn'], inplace=True)
data_fun.head(10)

id announcedOn  \
120250  00003f34-ac89-4090-b592-30f2b70ff5a4  2018-08-15   
54720   00014e45-238c-4ce1-a9d1-de7c6d5bbf2c  2018-07-09   
54719   00014e45-238c-4ce1-a9d1-de7c6d5bbf2c  2020-03-27   
79756   0001b1aa-4a8b-4ee2-b4c0-6bce0318c2c1  2022-10-31   
253767  0003088d-8c0b-46f6-910a-6dea65777ac4  2018-05-15   
303969  0004d27e-6d98-4f29-a708-b989e3e38f81  2017-12-01   
31069   0004e479-1be1-49d9-b54d-e6c2085dc18e  2020-01-01   
31070   0004e479-1be1-49d9-b54d-e6c2085dc18e  2020-05-15   
31071   0004e479-1be1-49d9-b54d-e6c2085dc18e  2021-06-15   
31072   0004e479-1be1-49d9-b54d-e6c2085dc18e  2021-06-18   

                                              moneyRaised  
120250  {"amount": 15098200, "currency": "USD", "amoun...  
54720   {"amount": 11814500, "currency": "USD", "amoun...  
54719   {"amount": 27167600, "currency": "USD", "amoun...  
79756   {"amount": 5000000, "currency": "USD", "amount...  
253767  {"amount": 2804700, "currency": "USD", "amount...  
303969                                                NaN  
31069   {"amount": 22500000, "currency": "USD", "amoun...  
31070   {"amount": 2500000, "currency": "USD", "amount...  
31071   {"amount": 25000000, "currency": "USD", "amoun...  
31072   {"amount": 5000000, "currency": "USD", "amount...

In [422]:
from preprocessing.format_moneyRaised import format_moneyRaised

format_moneyRaised(data_fun)

id announcedOn  \
120250  00003f34-ac89-4090-b592-30f2b70ff5a4  2018-08-15   
54720   00014e45-238c-4ce1-a9d1-de7c6d5bbf2c  2018-07-09   
54719   00014e45-238c-4ce1-a9d1-de7c6d5bbf2c  2020-03-27   
79756   0001b1aa-4a8b-4ee2-b4c0-6bce0318c2c1  2022-10-31   
253767  0003088d-8c0b-46f6-910a-6dea65777ac4  2018-05-15   
...                                      ...         ...   
296908  fffe3abf-f785-4876-9281-aa45cd70a180  2016-05-01   
296907  fffe3abf-f785-4876-9281-aa45cd70a180  2016-12-01   
74457   ffff528c-dd2d-489a-b6ca-1956a526b80a  2020-01-02   
306647  ffffa984-a16c-4942-a968-1c5806641c30  2018-06-01   
306648  ffffa984-a16c-4942-a968-1c5806641c30  2018-09-07   

                                              moneyRaised  moneyRaised_USD  
120250  {"amount": 15098200, "currency": "USD", "amoun...       15098200.0  
54720   {"amount": 11814500, "currency": "USD", "amoun...       11814500.0  
54719   {"amount": 27167600, "currency": "USD", "amoun...       27167600.0  
79756   {"amount": 5000000, "currency": "USD", "amount...        5000000.0  
253767  {"amount": 2804700, "currency": "USD", "amount...        2804700.0  
...                                                   ...              ...  
296908  {"amount": 46340300, "currency": "USD", "amoun...       46340300.0  
296907  {"amount": 116184300, "currency": "USD", "amou...      116184300.0  
74457   {"amount": 300000000, "currency": "USD", "amou...      300000000.0  
306647  {"amount": 12000000, "currency": "USD", "amoun...       12000000.0  
306648  {"amount": 50000000, "currency": "USD", "amoun...       50000000.0  

[123450 rows x 4 columns]

In [423]:
data_fun['moneyRaised_USD'] = data_fun['moneyRaised_USD']/100

In [424]:
data_fun

id announcedOn  \
120250  00003f34-ac89-4090-b592-30f2b70ff5a4  2018-08-15   
54720   00014e45-238c-4ce1-a9d1-de7c6d5bbf2c  2018-07-09   
54719   00014e45-238c-4ce1-a9d1-de7c6d5bbf2c  2020-03-27   
79756   0001b1aa-4a8b-4ee2-b4c0-6bce0318c2c1  2022-10-31   
253767  0003088d-8c0b-46f6-910a-6dea65777ac4  2018-05-15   
...                                      ...         ...   
296908  fffe3abf-f785-4876-9281-aa45cd70a180  2016-05-01   
296907  fffe3abf-f785-4876-9281-aa45cd70a180  2016-12-01   
74457   ffff528c-dd2d-489a-b6ca-1956a526b80a  2020-01-02   
306647  ffffa984-a16c-4942-a968-1c5806641c30  2018-06-01   
306648  ffffa984-a16c-4942-a968-1c5806641c30  2018-09-07   

                                              moneyRaised  moneyRaised_USD  
120250  {"amount": 15098200, "currency": "USD", "amoun...         150982.0  
54720   {"amount": 11814500, "currency": "USD", "amoun...         118145.0  
54719   {"amount": 27167600, "currency": "USD", "amoun...         271676.0  
79756   {"amount": 5000000, "currency": "USD", "amount...          50000.0  
253767  {"amount": 2804700, "currency": "USD", "amount...          28047.0  
...                                                   ...              ...  
296908  {"amount": 46340300, "currency": "USD", "amoun...         463403.0  
296907  {"amount": 116184300, "currency": "USD", "amou...        1161843.0  
74457   {"amount": 300000000, "currency": "USD", "amou...        3000000.0  
306647  {"amount": 12000000, "currency": "USD", "amoun...         120000.0  
306648  {"amount": 50000000, "currency": "USD", "amoun...         500000.0  

[123450 rows x 4 columns]

In [425]:
data_fun.drop(columns=['moneyRaised', 'announcedOn'], inplace=True)
data_fun = data_fun[data_fun['moneyRaised_USD'] != 0]
data_fun.head(20)

id  moneyRaised_USD
120250  00003f34-ac89-4090-b592-30f2b70ff5a4         150982.0
54720   00014e45-238c-4ce1-a9d1-de7c6d5bbf2c         118145.0
54719   00014e45-238c-4ce1-a9d1-de7c6d5bbf2c         271676.0
79756   0001b1aa-4a8b-4ee2-b4c0-6bce0318c2c1          50000.0
253767  0003088d-8c0b-46f6-910a-6dea65777ac4          28047.0
31069   0004e479-1be1-49d9-b54d-e6c2085dc18e         225000.0
31070   0004e479-1be1-49d9-b54d-e6c2085dc18e          25000.0
31071   0004e479-1be1-49d9-b54d-e6c2085dc18e         250000.0
31072   0004e479-1be1-49d9-b54d-e6c2085dc18e          50000.0
31074   0004e479-1be1-49d9-b54d-e6c2085dc18e        2200000.0
16337   0006b408-8356-4b54-a527-c256655e0c5b        3750000.0
302593  0007bff2-46fa-4e80-b815-45a05569e9d7         343358.0
302594  0007bff2-46fa-4e80-b815-45a05569e9d7        9940530.0
302592  0007bff2-46fa-4e80-b815-45a05569e9d7       12096827.0
32382   00080c23-0510-4f26-9010-9faabe2efc3e        2000000.0
184108  00095129-2085-4236-abca-e89d6b7a0740         940320.0
77032   000b7e41-7dd9-4941-8217-cddef1ae8a09         500000.0
237870  000c45ec-0afd-4b8e-a4bd-d1bc2396690a         485000.0
6446    000ff03c-fa33-482d-80ff-844d99a2f507       15000000.0
6447    000ff03c-fa33-482d-80ff-844d99a2f507       15000000.0

In [426]:
data_fun['index'] = data_fun.groupby('id').cumcount() + 1
# pivot the data to create the desired output
output = data_fun.pivot(index='id', columns='index', values='moneyRaised_USD')
# rename the columns to match the desired output
output.columns = [str(i) for i in output.columns]
data_fun = output

/var/folders/rt/gch51dj925v29nlt01bkgkg80000gn/T/ipykernel_71457/479812745.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data_fun['index'] = data_fun.groupby('id').cumcount() + 1


In [427]:
data_fun.reset_index(inplace=True)
data_fun = data_fun.iloc[:, 0:6]
data_fun

id          1          2         3  \
0      00003f34-ac89-4090-b592-30f2b70ff5a4   150982.0        NaN       NaN   
1      00014e45-238c-4ce1-a9d1-de7c6d5bbf2c   118145.0   271676.0       NaN   
2      0001b1aa-4a8b-4ee2-b4c0-6bce0318c2c1    50000.0        NaN       NaN   
3      0003088d-8c0b-46f6-910a-6dea65777ac4    28047.0        NaN       NaN   
4      0004e479-1be1-49d9-b54d-e6c2085dc18e   225000.0    25000.0  250000.0   
...                                     ...        ...        ...       ...   
42699  fffd4279-32ff-46d3-8b63-58eb1aead505   250000.0        NaN       NaN   
42700  fffdd79a-655f-4d82-b09a-68e81d98ef90   131207.0   333078.0       NaN   
42701  fffe3abf-f785-4876-9281-aa45cd70a180   463403.0  1161843.0       NaN   
42702  ffff528c-dd2d-489a-b6ca-1956a526b80a  3000000.0        NaN       NaN   
42703  ffffa984-a16c-4942-a968-1c5806641c30   120000.0   500000.0       NaN   

             4          5  
0          NaN        NaN  
1          NaN        NaN  
2          NaN        NaN  
3          NaN        NaN  
4      50000.0  2200000.0  
...        ...        ...  
42699      NaN        NaN  
42700      NaN        NaN  
42701      NaN        NaN  
42702      NaN        NaN  
42703      NaN        NaN  

[42704 rows x 6 columns]

In [428]:
for i in range(1, 6):
    data_fun.rename(columns={str(i): f'Round {str(i)}'}, inplace=True)

data_fun

id    Round 1    Round 2   Round 3  \
0      00003f34-ac89-4090-b592-30f2b70ff5a4   150982.0        NaN       NaN   
1      00014e45-238c-4ce1-a9d1-de7c6d5bbf2c   118145.0   271676.0       NaN   
2      0001b1aa-4a8b-4ee2-b4c0-6bce0318c2c1    50000.0        NaN       NaN   
3      0003088d-8c0b-46f6-910a-6dea65777ac4    28047.0        NaN       NaN   
4      0004e479-1be1-49d9-b54d-e6c2085dc18e   225000.0    25000.0  250000.0   
...                                     ...        ...        ...       ...   
42699  fffd4279-32ff-46d3-8b63-58eb1aead505   250000.0        NaN       NaN   
42700  fffdd79a-655f-4d82-b09a-68e81d98ef90   131207.0   333078.0       NaN   
42701  fffe3abf-f785-4876-9281-aa45cd70a180   463403.0  1161843.0       NaN   
42702  ffff528c-dd2d-489a-b6ca-1956a526b80a  3000000.0        NaN       NaN   
42703  ffffa984-a16c-4942-a968-1c5806641c30   120000.0   500000.0       NaN   

       Round 4    Round 5  
0          NaN        NaN  
1          NaN        NaN  
2          NaN        NaN  
3          NaN        NaN  
4      50000.0  2200000.0  
...        ...        ...  
42699      NaN        NaN  
42700      NaN        NaN  
42701      NaN        NaN  
42702      NaN        NaN  
42703      NaN        NaN  

[42704 rows x 6 columns]

## Join `data_ind`, `data_tec`, and `data_fun` to original dataset

In [463]:
data.columns

Index(['id', 'name', 'website', 'short_description', 'ipo_status',
       'founded_on', 'went_public_on', 'exited_on', 'num_funding_rounds',
       'last_equity_funding_type', 'last_equity_funding_total',
       'last_funding_at', 'headquartersCountry', 'headquartersRegion',
       'employeeCount', 'industry_name', 'technology_name', 'announcedOn',
       'moneyRaised'],
      dtype='object')

In [464]:
data_merged = data.drop(columns=['industry_name',
                                 'technology_name',
                                 'announcedOn',
                                 'moneyRaised'])

In [465]:
data_merged = data_merged.drop_duplicates().reset_index(drop=True)

In [466]:
data_merged.shape

(56667, 15)

### Merging Industry:

In [469]:
data_ind.shape

(56667, 41)

In [470]:
data_ind.head(2)

id  Advertising  Agriculture and Farming  \
0  00003f34-ac89-4090-b592-30f2b70ff5a4          0.0                      0.0   
1  00014e45-238c-4ce1-a9d1-de7c6d5bbf2c          0.0                      0.0   

   Clothing and Apparel  Commerce and Shopping  Community and Lifestyle  \
0                   0.0                    0.0                      0.0   
1                   0.0                    0.0                      1.0   

   Computer Hardware  Consumer Electronics  Consumer Goods  \
0                0.0                   0.0             0.0   
1                0.0                   0.0             0.0   

   Content and Publishing  ...  Payments  Privacy and Security  \
0                     0.0  ...       0.0                   0.0   
1                     0.0  ...       0.0                   0.0   

   Professional Services  Real Estate and Construction  Sales and Marketing  \
0                    0.0                           0.0                  0.0   
1                    0.0                           0.0                  0.0   

   Software  Sports  Transportation  Travel and Tourism  Video  
0       0.0     0.0             0.0                 0.0    0.0  
1       0.0     0.0             0.0                 0.0    0.0  

[2 rows x 41 columns]

In [471]:
data_merged_ind = data_merged.merge(data_ind, on='id', how='left')
data_merged_ind.shape

(56667, 55)

### Merging Technology:

In [472]:
data_tec.shape

(56667, 9)

In [473]:
data_tec.head(2)

id  AR and VR  Artificial Intelligence  \
0  00003f34-ac89-4090-b592-30f2b70ff5a4        0.0                      0.0   
1  00014e45-238c-4ce1-a9d1-de7c6d5bbf2c        0.0                      0.0   

   Biotechnology  BlockChain  Hardware  Science and Engineering  Software  \
0            0.0         0.0       0.0                      0.0       0.0   
1            0.0         0.0       0.0                      0.0       1.0   

   Sustainability  
0             1.0  
1             0.0

In [474]:
data_merged_tec = data_merged_ind.merge(data_tec, on='id', how='left')
data_merged_tec.shape

(56667, 63)

### Merging Funding Rounds:

In [476]:
data_fun.shape

(42704, 6)

In [477]:
data_fun.head(2)

id   Round 1   Round 2  Round 3  Round 4  \
0  00003f34-ac89-4090-b592-30f2b70ff5a4  150982.0       NaN      NaN      NaN   
1  00014e45-238c-4ce1-a9d1-de7c6d5bbf2c  118145.0  271676.0      NaN      NaN   

   Round 5  
0      NaN  
1      NaN

In [478]:
data_merged_all = data_merged_tec.merge(data_fun, on='id', how='left')
data_merged_all.shape

(56667, 68)

# Final Dataset post-duplicates

In [479]:
data_merged_all

id               name  \
0      75139f4c-6d9d-4a03-8b72-61f73f7be74a       Youcan Robot   
1      441e5d33-140e-40c6-ab5d-35f65ab8feed  Make it Home Safe   
2      59b8a226-eb24-4e7b-851d-924e0385d01c              Biveo   
3      f786e92e-7ac1-4035-9644-de30b8d8b6a5           Spaced-X   
4      45fff4cb-2c30-4fef-b8cd-c18239c9e6a0          Superflex   
...                                     ...                ...   
56662  86e5c2b7-1508-4bb6-b6b0-92269a93d0c6        TradeWindow   
56663  ed1d84d8-4072-459a-80ed-a18ec7a13513             Alcide   
56664  e3e6968e-0bd8-4fa8-b529-8fe87d208be9         Anvert Kft   
56665  2b2923fe-132d-4167-a97f-682e575df22f  Drōv Technologies   
56666  41b059a5-3b39-42a7-b13d-d7d3e1f51a61            Genster   

                                 website  \
0           https://www.youcanrobot.com/   
1            https://makeithomesafe.com/   
2                     https://biveo.com/   
3               http://www.spaced-x.com/   
4       http://superflextechnologies.com   
...                                  ...   
56662            https://tradewindow.io/   
56663               http://www.alcide.io   
56664                https://anvert.com/   
56665  https://www.drovtechnologies.com/   
56666  https://www.genster.cz/index.html   

                                       short_description ipo_status  \
0      Youcan Robot is an integrated company in the w...    private   
1      Make it Home Safe is a mobile app that provide...    private   
2          Business Video Production Directory and Guide    private   
3      Medical grade consumption accessories for the ...    private   
4      Apparel enhancing muscle strength targeting el...    private   
...                                                  ...        ...   
56662  TradeWindow is a global trade platform for exp...    private   
56663  Alcide provides a cloud-native security platfo...    private   
56664  Anvert is a web-based app that reduces the tim...    private   
56665  Drōv is a technology company developing innova...    private   
56666  Genster is a first shopping and styling servic...    private   

       founded_on went_public_on   exited_on  num_funding_rounds  \
0      2016-01-01            NaN         NaN                   1   
1      2017-01-01            NaN         NaN                   2   
2      2016-07-11            NaN         NaN                   1   
3      2018-02-24            NaN         NaN                   1   
4      2016-02-01            NaN         NaN                   2   
...           ...            ...         ...                 ...   
56662  2018-12-03            NaN         NaN                   3   
56663  2016-01-01            NaN  2021-01-31                   4   
56664  2018-01-01            NaN         NaN                   4   
56665  2016-01-01            NaN         NaN                   1   
56666  2018-01-01            NaN         NaN                   3   

      last_equity_funding_type  ... BlockChain Hardware  \
0                         seed  ...        0.0      1.0   
1                         seed  ...        0.0      0.0   
2                         seed  ...        0.0      0.0   
3                     pre_seed  ...        0.0      0.0   
4                          NaN  ...        0.0      0.0   
...                        ...  ...        ...      ...   
56662                 series_b  ...        0.0      0.0   
56663                 series_a  ...        0.0      0.0   
56664                     seed  ...        0.0      0.0   
56665                     seed  ...        0.0      1.0   
56666                     seed  ...        0.0      0.0   

      Science and Engineering Software_y  Sustainability    Round 1  \
0                         1.0        0.0             0.0        NaN   
1                         0.0        1.0             0.0    50000.0   
2                         0.0        1.0             0.0   500000.0   
3                         0.0        0.0             0.0

In [483]:
data_merged_all.describe()

num_funding_rounds  employeeCount   Advertising  \
count        56667.000000   52234.000000  56667.000000   
mean             2.001941      57.874201      0.028094   
std              1.572155     449.148989      0.165243   
min              1.000000       0.000000      0.000000   
25%              1.000000       6.000000      0.000000   
50%              1.000000       6.000000      0.000000   
75%              2.000000      31.000000      0.000000   
max             22.000000   32098.000000      1.000000   

       Agriculture and Farming  Clothing and Apparel  Commerce and Shopping  \
count             56667.000000          56667.000000           56667.000000   
mean                  0.026806              0.019341               0.105758   
std                   0.161517              0.137722               0.307531   
min                   0.000000              0.000000               0.000000   
25%                   0.000000              0.000000               0.000000   
50%                   0.000000              0.000000               0.000000   
75%                   0.000000              0.000000               0.000000   
max                   1.000000              1.000000               1.000000   

       Community and Lifestyle  Computer Hardware  Consumer Electronics  \
count              56667.00000       56667.000000          56667.000000   
mean                   0.02467           0.137311              0.069811   
std                    0.15512           0.344178              0.254831   
min                    0.00000           0.000000              0.000000   
25%                    0.00000           0.000000              0.000000   
50%                    0.00000           0.000000              0.000000   
75%                    0.00000           0.000000              0.000000   
max                    1.00000           1.000000              1.000000   

       Consumer Goods  ...    BlockChain      Hardware  \
count    56667.000000  ...  56667.000000  56667.000000   
mean         0.014576  ...      0.048547      0.067270   
std          0.119851  ...      0.214920      0.250492   
min          0.000000  ...      0.000000      0.000000   
25%          0.000000  ...      0.000000      0.000000   
50%          0.000000  ...      0.000000      0.000000   
75%          0.000000  ...      0.000000      0.000000   
max          1.000000  ...      1.000000      1.000000   

       Science and Engineering    Software_y  Sustainability       Round 1  \
count             56667.000000  56667.000000    56667.000000  4.270400e+04   
mean                  0.026576      0.643743        0.060211  7.603390e+06   
std                   0.160843      0.478897        0.237880  1.171411e+08   
min                   0.000000      0.000000        0.000000  1.000000e+00   
25%                   0.000000      0.000000        0.000000  1.396152e+05   
50%                   0.000000      1.000000        0.000000  5.652885e+05   
75%                   0.000000      1.000000        0.000000  2.430765e+06   
max                   1.000000      1.000000        1.000000  1.660000e+10   

            Round 2       Round 3       Round 4       Round 5  
count  1.990500e+04  1.088300e+04  6.201000e+03  3.560000e+03  
mean   9.656692e+06  1.334077e+07  1.749175e+07  2.202715e+07  
std    8.165809e+07  5.725938e+07  5.928507e+07  7.050248e+07  
min    1.000000e+00  2.750000e+02  2.770000e+02  3.730000e+02  
25%    2.650000e+05  4.169455e+05  5.650000e+05  6.500000e+05  
50%    1.295000e+06  2.000000e+06  2.722234e+06  3.346794e+06  
75%    5.000000e+06  8.399292e+06  1.200000e+07  1.553188e+07  
max    9.049262e+09  2.262315e+09  2.600000e+09  2.400000e+09  

[8 rows x 55 columns]

In [484]:
data_merged_all.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 56667 entries, 0 to 56666
Data columns (total 68 columns):
 #   Column                            Non-Null Count  Dtype  
---  ------                            --------------  -----  
 0   id                                56667 non-null  object 
 1   name                              56667 non-null  object 
 2   website                           55204 non-null  object 
 3   short_description                 56665 non-null  object 
 4   ipo_status                        56667 non-null  object 
 5   founded_on                        56667 non-null  object 
 6   went_public_on                    664 non-null    object 
 7   exited_on                         3325 non-null   object 
 8   num_funding_rounds                56667 non-null  int64  
 9   last_equity_funding_type          48618 non-null  object 
 10  last_equity_funding_total         35500 non-null  object 
 11  last_funding_at                   56667 non-null  object 
 12  head